src/harness/reference_models/examples/entities.py

```diff
@@ -116,24 +123,92 @@ def GenerateCbsdList(n_cbsd, template_cbsd,
-        cbsds.append(cbsd)
-    else:
+    for azimuth in azimuths:
       cbsd = Cbsd(lat, lng,
-                  t.height_agl, t.is_indoor, t.category, t.eirp_dbm,
-                  np.random.uniform(0, 360, 1)[0], t.antenna_beamwidth)
+                  t.height_agl, t.is_indoor, t.category, t.eirp_dbm_mhz,
+                  int(azimuth + np.random.uniform(0, 360)) % 360,
+                  t.antenna_beamwidth, t.antenna_gain)
       cbsds.append(cbsd)
 
   return cbsds
+
+
+def GenerateCbsdsInPolygon(num_cbsds, template_cbsd, polygon,
+                           nlcd_driver=None, urban_areas=None):
+  """Returns an inland Cbsd within polygon, possibly in urban area.
+
+
+  Args:
+    template_cbsd: An |Cbsd| used as a template.
+    polygon: A |shapely.Polygon| where to put the Cbsd.
+    nlcd_driver: The land cover driver for detecting inland positions (optional).
+    urban_areas: An optional  |shapely.MultiPolygon| defining the urban areas.
+      If not specified, do not enforce the Cbsd to be in urban areas.
+  """
+  t = template_cbsd
+  bounds = polygon.bounds
+  cbsds = []
+  while len(cbsds) < num_cbsds:
+    while True:
+      lng = np.random.uniform(bounds[0], bounds[2])
+      lat = np.random.uniform(bounds[1], bounds[3])
+      if not sgeo.Point(lng, lat).within(polygon):
+        continue
+      if nlcd_driver is not None:
+        land_cover = nlcd_driver.GetLandCoverCodes(lat, lng)
+        if land_cover <= nlcd.LandCoverCodes.PERENNIAL_SNOW:
+          continue
+
+      if urban_areas is not None:
+        if not sgeo.Point(lng, lat).within(urban_areas):
+          continue
+      break
+
+    azimuth_offset = np.random.uniform(0, 360)
+    azimuths = (t.antenna_azimuth if isinstance(t.antenna_azimuth, list)
+                else [t.antenna_azimuth])
+    for azimuth in azimuths:
+      cbsds.append(Cbsd(lat, lng, t.height_agl, t.is_indoor, t.category,
+                        t.eirp_dbm_mhz, azimuth,
+                        t.antenna_beamwidth, t.antenna_gain))
+
+  return cbsds
+
+
+# Utility routines to convert Cbsd object into registration and grant requests.
+def GetCbsdRegistrationRequest(cbsd):
+  """Returns a CBSD registration request from the |Cbsd| object.
+  """
+  return {
+      'cbsdCategory': cbsd.category,
+      'installationParam': {
+          'latitude': cbsd.latitude,
+          'longitude': cbsd.longitude,
+          'height': cbsd.height_agl,
+          'heightType': 'AGL',
+          'indoorDeployment': cbsd.is_indoor,
+          'antennaBeamwidth': cbsd.antenna_beamwidth,
+          'antennaAzimuth': cbsd.antenna_azimuth,
+          'antennaGain': cbsd.antenna_gain,
+      }
+  }
+
+
+def GetCbsdGrantRequest(cbsd, min_freq_mhz, max_freq_mhz):
+  """Returns a CBSD grant request from a |Cbsd| object and a min/max frequency.
+  """
+  return {
+      'operationParam': {
+          'operationFrequencyRange': {
+              'lowFrequency': min_freq_mhz * 1e6,
```

Comment:
I think we need to typecast this to int as multiplying it with 1e6 will generate decimal.

Reply:
Yes it is decimal, I dont think the reference models cares, but for sake of generality if we use this purely test code to test/benchmark other pieces which care about that.